src/main/java/org/apache/sysds/runtime/instructions/fed/MatrixIndexingFEDInstruction.java

```diff
@@ -0,0 +1,196 @@
+						fedMapping.getFederatedRanges()[i].setEndDim(1, fedMapping.getFederatedRanges()[i].getBeginDims()[1]);
+						rsn = -1;
+						ren = rsn;
+						csn = rsn;
+						cen = rsn;
+					}
+					break;
+				case OTHER:
+					rsn = (ixrange.rowStart >= rs && ixrange.rowStart < re) ? (ixrange.rowStart - rs) : 0;
+					ren = (ixrange.rowEnd >= rs && ixrange.rowEnd < re) ? (ixrange.rowEnd - rs) : (re - rs - 1);
+					csn = (ixrange.colStart >= cs && ixrange.colStart < ce) ? (ixrange.colStart - cs) : 0;
+					cen = (ixrange.colEnd >= cs && ixrange.colEnd < ce) ? (ixrange.colEnd - cs) : (ce - cs - 1);
+
+					if((ixrange.colStart < ce) && (ixrange.colEnd >= cs) && (ixrange.rowStart < re) && (ixrange.rowEnd >= rs)) {
+						if (fedMapping.getFederatedRanges()[i-1].getSize() != 0 || lastEndDim[0] + lastEndDim[1] == 0) {
+							fedMapping.getFederatedRanges()[i].setBeginDim(0, i != 0 ? fedMapping.getFederatedRanges()[i - 1].getEndDims()[0] : 0);
+							fedMapping.getFederatedRanges()[i].setBeginDim(1, i != 0 ? fedMapping.getFederatedRanges()[i - 1].getEndDims()[1] : 0);
+							fedMapping.getFederatedRanges()[i].setEndDim(0, ren - rsn + 1 + (i == 0 ? 0 : fedMapping.getFederatedRanges()[i - 1].getEndDims()[0]));
+							fedMapping.getFederatedRanges()[i].setEndDim(1, cen - csn + 1 + (i == 0 ? 0 : fedMapping.getFederatedRanges()[i - 1].getEndDims()[1]));
+						} else {
+							fedMapping.getFederatedRanges()[i].setBeginDim(0, 1 + ren - rsn + lastBeginDim[0]);
+							fedMapping.getFederatedRanges()[i].setBeginDim(1, cen - csn + lastBeginDim[1]);
+							fedMapping.getFederatedRanges()[i].setEndDim(0, 1 + ren - rsn + lastEndDim[0]);
+							fedMapping.getFederatedRanges()[i].setEndDim(1, cen - csn + lastEndDim[1]);
+						}
+						lastEndDim = fedMapping.getFederatedRanges()[i].getEndDims();
+						lastBeginDim = fedMapping.getFederatedRanges()[i].getBeginDims();
+					}
+					else {
+						fedMapping.getFederatedRanges()[i].setBeginDim(0, i != 0 ? fedMapping.getFederatedRanges()[i - 1].getEndDims()[0] : 0);
+						fedMapping.getFederatedRanges()[i].setEndDim(0, fedMapping.getFederatedRanges()[i].getBeginDims()[0]);
+
+						fedMapping.getFederatedRanges()[i].setBeginDim(1, i != 0 ? fedMapping.getFederatedRanges()[i - 1].getEndDims()[1] : 0);
+						fedMapping.getFederatedRanges()[i].setEndDim(1, fedMapping.getFederatedRanges()[i].getBeginDims()[1]);
+
+						rsn = -1;
+						ren = rsn;
+						csn = rsn;
+						cen = rsn;
+					}
+			}
+			ixs.put(fedMapping.getFederatedRanges()[i].toString(), new IndexRange(rsn, ren, csn, cen));
+		}
+//		List<MatrixBlock> mb = new ArrayList<>();
+		long varID = FederationUtils.getNextFedDataID();
+		FederationMap sortedMapping = fedMapping.mapParallel(varID, (range, data) -> {
+			try {
+				FederatedResponse response = data
+					.executeFederatedOperation(new FederatedRequest(FederatedRequest.RequestType.EXEC_UDF, -1,
+						new SliceMatrix(data.getVarID(), varID, ixs.get(range.toString())))).get();
```

Comment:
I think you can use the Federated Range object it self to look up the entries, instead of the to string of it.